x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_logic.ts

```diff
@@ -0,0 +1,263 @@
+    setTokenType: (tokenType) => tokenType,
+    toggleCredentialsForm: (apiToken = { ...defaultApiToken }) => apiToken,
+    hideCredentialsForm: false,
+    resetCredentials: false,
+    initializeCredentialsData: true,
+    fetchCredentials: (page) => page,
+    fetchDetails: true,
+    deleteApiKey: (tokenName) => tokenName,
+  }),
+  reducers: () => ({
+    apiTokens: [
+      [],
+      {
+        setCredentialsData: (_, { apiTokens }) => apiTokens,
+        onApiTokenCreateSuccess: (apiTokens, apiToken) => [...apiTokens, apiToken],
+        onApiTokenUpdateSuccess: (apiTokens, apiToken) => [
+          ...apiTokens.filter((token) => token.name !== apiToken.name),
+          apiToken,
+        ],
+        onApiKeyDelete: (apiTokens, tokenName) =>
+          apiTokens.filter((token) => token.name !== tokenName),
+      },
+    ],
+    meta: [
+      {},
+      {
+        setCredentialsData: (_, { meta }) => meta,
+      },
+    ],
+    isCredentialsDetailsComplete: [
+      false,
+      {
+        setCredentialsDetails: () => true,
+        resetCredentials: () => false,
+      },
+    ],
+    isCredentialsDataComplete: [
+      false,
+      {
+        setCredentialsData: () => true,
+        resetCredentials: () => false,
+      },
+    ],
+    engines: [
+      [],
+      {
+        setCredentialsDetails: (_, { engines }) => engines,
+      },
+    ],
+    nameInputBlurred: [
+      false,
+      {
+        setNameInputBlurred: (_, nameInputBlurred) => nameInputBlurred,
+      },
+    ],
+    activeApiToken: [
+      defaultApiToken,
+      {
+        addEngineName: (activeApiToken, engineName) => ({
+          ...activeApiToken,
+          engines: [...(activeApiToken.engines || []), engineName],
+        }),
+        removeEngineName: (activeApiToken, engineName) => ({
+          ...activeApiToken,
+          engines: (activeApiToken.engines || []).filter((name) => name !== engineName),
+        }),
+        setAccessAllEngines: (activeApiToken, accessAll) => ({
+          ...activeApiToken,
+          access_all_engines: accessAll,
+          engines: accessAll ? [] : activeApiToken.engines,
+        }),
+        onApiTokenCreateSuccess: () => defaultApiToken,
+        onApiTokenUpdateSuccess: () => defaultApiToken,
+        setTokenName: (activeApiToken, name) => ({ ...activeApiToken, name: formatApiName(name) }),
+        setTokenReadWrite: (activeApiToken, { name, checked }) => ({
+          ...activeApiToken,
+          [name]: checked,
+        }),
+        setTokenType: (activeApiToken, tokenType) => ({
+          ...activeApiToken,
+          access_all_engines: tokenType === ADMIN ? false : activeApiToken.access_all_engines,
+          engines: tokenType === ADMIN ? [] : activeApiToken.engines,
+          write: tokenType === PRIVATE,
+          read: tokenType === PRIVATE,
+          type: tokenType,
+        }),
+        toggleCredentialsForm: (_, activeApiToken) => activeApiToken || defaultApiToken,
```

Comment:
So, this is the last uncovered branch line in the file... and I'd like to make a case for actually getting rid of the `|| defaultApiToken` fallback:

<img width="746" alt="" src="https://user-images.githubusercontent.com/549407/94050885-b78d6580-fd8b-11ea-9558-ac854d20b27a.png">

I was actually super surprised that this wasn't covered by

https://github.com/elastic/kibana/blob/175198d5c1eb29f2c42c7adedf7098f57b433516/x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_logic.test.ts#L899-L907

but upon closer inspection, the issue is this _earlier_ fallback in our `actions` above:

<img width="640" alt="" src="https://user-images.githubusercontent.com/549407/94051209-118e2b00-fd8c-11ea-8ef6-ff912052dab7.png">

Since `(param = default)` fallbacks only catch `undefined`, in theory we could still trigger the `||` falsey catch by passing `null`, `0`, or `''` etc (I don't know why we'd ever do so in real life, but it's a theory). I tried this in a test like so:

```
CredentialsLogic.actions.toggleCredentialsForm(null);
```

... However, this causes another error in another `toggleCredentialsForm` reducer that expects an obj / destructures:

<img width="788" alt="" src="https://user-images.githubusercontent.com/549407/94051532-81041a80-fd8c-11ea-8b33-e4eb5a01b387.png">

Which leads me to conclude we're never going to practically use the `||` fallback over the earlier `= {}` default param and we should just nuke it. Yay/nay? 😃 

```suggestion
        toggleCredentialsForm: (_, activeApiToken) => activeApiToken,
```

Reply:
Great catch.

Reply:
https://github.com/elastic/kibana/pull/77626/commits/5a46c0845987e0716f48defc8e6f5114b20e3b0c